### ISSUE TITLE
Add compiler-internal lints

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#49509]
There are various "patterns" we don't want in the compiler out of one reason or another. I'm using this issue to collect them and create lints for detecting those patterns. The lints will only be available while building the compiler via a -Z flag. The flag will be enabled by the bootstrap script, so we don't have to worry about accidentally running them for users.

* [ ] forbid uses of `HashSet` or `HashMap`
* [ ] suggest using `TyCtxtAt` over `TyCtxt` + `Span` as arguments

**Comment by leonardo-m** [comment]
Why are HashSet and HashMap forbidden as dog food for the compiler? And why aren't they "forbidden" for user code? This seems a hint.

**Comment by Ixrec** [comment]
If I remember correctly, it's because the standard HashSet/Map make a significant effort to protect against DoS attacks, i.e. they try to make it impossible for an attacker to induce huge performance cliffs with special sequences of malicious inputs (for example: https://accidentallyquadratic.tumblr.com/post/153545455987/rust-hash-iteration-reinsertion), but that concern doesn't apply inside the compiler so it makes more sense to use a faster implementation that does not offer any such protection.

**Comment by eddyb** [comment]
@Ixrec Also, it's useful to have deterministic compilation, without sources of randomness.

**Comment by oli-obk** [comment]
cc https://github.com/rust-lang/rust-clippy/issues/3724

Implementing this lint in clippy will make rustc miss out on it. On the other hand, we could setup clippy to run on rustc similar to tidy

**Comment by flip1995** [comment]
The first and the last two items can be checked.

About the `Clone` on `Copy` lint: Would it make sense to uplift this lint completely to rustc? What's the state of the "no new lints in rustc" rule?

**Comment by oli-obk** [comment]
That rule still applies, can we move it to rustc as an internal lint but have clippy "reexport" it as its own lint?

**Comment by flip1995** [comment]
This could be possible by making the LintPass public in rustc and re-registering the LintPass in Clippy. I'm curious how this will work with `tool_lints` 🤔

**Comment by oli-obk** [comment]
adding new lint ideas

* don't use `TyS` outside `rustc::ty`
* don't use `Vec::new()` use `vec![]` (cc @eddyb)

**Comment by eddyb** [comment]
Hmm, instead of "don't do X outside of the Y module" - can we simply allow it, not on the whole Y module, but on the smallest piece of Y that needs to do X? E.g.:
```rust
#[allow(usage_of_ty_tykind)]
pub use self::TyKind::*;
```

**Comment by Centril** [comment]
New internal lint idea:
- ban identifiers `did` (also `vid`, ...) when they refer to `DefID` and similar.

cc @oli-obk https://github.com/rust-lang/rust/pull/61350#discussion_r289366939

**Comment by flip1995** [comment]
Internal lint idea:
* Check that error / lint / help / ... messages start lower case

This is a hard to find (I think I saw it documented somewhere :thinking:) rule of the rust compiler, which new contributors, especially in Clippy often get wrong (and is sometimes overlooked by the reviewers). Since rustc has a huge amount of functions that can produce error / lint / help / ... messages, this lint could be a bit tedious to implement.

**Comment by oli-obk** [comment]
Moderately stupid idea: just panic in the `Diagnostic` type if a string argument starts with a capitalized character.

**Comment by Centril** [comment]
cc https://github.com/rust-lang/rust/issues/63186

**Comment by jyn514** [comment]
>  ban pub use foo::bar;, #65233

This issue was closed, I think it should be removed from the list.

>  calling clone on Copy types (#27266)

That issue was closed because there's a clippy lint for it - maybe bootstrap should just pass `-W clippy::clone_on_copy`?

**Comment by oli-obk** [comment]
> That issue was closed because there's a clippy lint for it - maybe bootstrap should just pass -W clippy::clone_on_copy?

bootstrap doesn't run with clippy yet. I guess we could replace `./x.py check` with a clippy run with the beta clippy?

**Comment by jyn514** [comment]
> > That issue was closed because there's a clippy lint for it - maybe bootstrap should just pass -W clippy::clone_on_copy?
> 
> bootstrap doesn't run with clippy yet. I guess we could replace `./x.py check` with a clippy run with the beta clippy?

It works imperfectly since https://github.com/rust-lang/rust/pull/77351. It shouldn't be terribly hard to make it use beta clippy, I just haven't gotten around to it (right now it uses the default clippy of your host toolchain).

**Comment by oli-obk** [comment]
@nnethercote @Zalathar I assume you want the second point to be taken off this list then?

**Comment by Zalathar** [comment]
I don’t have much of an opinion on `TyCtxtAt` in general, because I don’t really know what it’s even supposed to be used for, and I don’t think I’ve ever been in a situation where I wanted to attach a single span of unknown significance to a TyCtxt.

So my irritation at having to deal with TyCtxtAt in hooks was that it was an extra hurdle to navigate, in contrast to queries, which never made me think about its existence at all.

**Comment by nnethercote** [comment]
I find `TyCtxtAt` confusing in general. I played around with removing it in favour of `TyCtxt` + explicit `Span`, but it's wired in too deeply in the query system to remove easily.

**Comment by oli-obk** [comment]
The query system uses it to report where a query fails. It can't be an argument to the query. The query body doesn't need it

**Comment by nnethercote** [comment]
Yeah, but there are a bunch of places in `rustc_const_eval` where `TyCxtAt` can(?) be replaced with `TyCtxt`+`Span`, that's what I was looking at before I realized how hardwired it is with queries.

**Comment by oli-obk** [comment]
That's not what we want tho. The span as an arg will not make a query cycle have nice spans.

What you're thinking of is the TyCtxtAt in the interpreter, which is just a convenience thing so we don't forget to `.at` before calling queries

**Comment by karolzwolak** [comment]
@oli-obk just wanted to make sure `don't use Vec::new() use vec![]` is still relevant, before I start working on it. Also is it worth looking into adding a `x clippy` command similar to `x tidy`?

**Comment by oli-obk** [comment]
I don't know. Seems contentious enough that we may just want to remove it from the list

**Comment by nnethercote** [comment]
Some grepping in `compiler/` finds:
- 878 occurrences of `Vec::new()`
- 716 occurrences of `vec![]`

Changing all the former to the latter doesn't seem like a win.
